File: 112. Path Sum/v1/pathSum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {boolean}
 */
var hasPathSum = function (root, targetSum) {
  if (!root) {
    return false;
  }
  return pathSumHelper(root, targetSum, 0);
};

function pathSumHelper(root, targetSum, currentSum) {
  if (!root) {
    return false;
  }
  currentSum = currentSum + root.val;
  if (!root.left && !root.right) {
    return targetSum === currentSum;
  }
  if (pathSumHelper(root.left, targetSum, currentSum)) {
    return true;
  }
  if (pathSumHelper(root.right, targetSum, currentSum)) {
    return true;
  }

  return false;
}

File: 112. Path Sum/v1/path-sum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func hasPathSum(root *TreeNode, targetSum int) bool {
	if root == nil {
		return false
	}
	return pathSumHelper(root, targetSum, 0)
}

func pathSumHelper(root *TreeNode, targetSum, currentSum int) bool {
	if root == nil {
		return false
	}

	currentSum += root.Val

	if root.Left == nil && root.Right == nil && currentSum == targetSum {
		return true
	}
	if pathSumHelper(root.Left, targetSum, currentSum) {
		return true
	}
	if pathSumHelper(root.Right, targetSum, currentSum) {
		return true
	}

	return false
}